rust/agents/relayer/src/relayer.rs

```diff
@@ -380,42 +373,54 @@ impl Relayer {
                 )
             })
             .collect();
-        let message_processor = MessageProcessor::new(
-            self.dbs.get(origin).unwrap().clone(),
-            self.whitelist.clone(),
-            self.blacklist.clone(),
-            metrics,
-            send_channels,
-            destination_ctxs,
-        );
 
-        let span = info_span!("MessageProcessor", origin=%message_processor.domain());
-        let processor = Processor::new(Box::new(message_processor));
+        let origin = origin.clone();
+        let db = self.dbs.get(&origin).unwrap().clone();
+        let whitelist = self.whitelist.clone();
+        let blacklist = self.blacklist.clone();
+        let core_metrics = self.core.metrics.clone();
+        let destination_chains = self.destination_chains.clone();
+        let span = info_span!("MessageProcessor", origin=%db.domain());
         tokio::spawn(async move {
-            let res = tokio::try_join!(processor.spawn())?;
-            info!(?res, "try_join finished for message processor");
-            Ok(())
+            loop {
+                let metrics =
+                    MessageProcessorMetrics::new(&core_metrics, &origin, destination_chains.keys());
+                let message_processor = MessageProcessor::new(
+                    db.clone(),
+                    whitelist.clone(),
+                    blacklist.clone(),
+                    metrics,
+                    send_channels.clone(),
+                    destination_ctxs.clone(),
+                );
+                let processor = Processor::new(Box::new(message_processor));
+                // the submitter runs forever, so we only need to handle error case
+                if let Err(err) = tokio::try_join!(processor.spawn()) {
```

Comment:
can you expand on what the goal is here? My understanding is that the changes made to processor.rs were to make it so that `spawn()` doesn't return a Result, so we'd only hit this condition here if the `spawn` task panics for some reason.

Do you have an opinion whether it's better to crash if one of these tasks panics or if we should try restarting `spawn`? We chatted a bit about this, I thought we settled on panicking. I'm more on the panicking side but I'm open to the respawning. But maybe I'm misunderstanding the logic here

The tradeoffs I see are:

Panicking pros:
- panics should happen extremely rarely and we probably want to learn about these, this would be more obvious to us if it does happen
- probably simpler to reason about task interactions

Panicking cons:
- liveness impact on other tasks

Re-spawning pros:
- no liveness impact on other tasks

Re-spawning cons:
- starting new tasks imo can lead to hard to reason about / buggy behavior due to the relationship that various tasks have with one another



To expand on the hard to reason about / buggy behavior:

- For the message processor tasks, if we restart it then the MessageProcessor starts from scratch and will send all the unprocessed messages from the db via the `send_channels` to the destination submitters again. So iiuc our destination submitter queues will have dupes of all the messages in them, which feels like it could have weird implications
- For the destination submitter tasks, if we restart one of those because of a panic, then we've lost the state of the queues that are populated by receiving unprocessed messages via the message processor's `send_channels`. So we'd only be able to process new messages that come in from the message processor and wouldn't process any older messages that the message processor has already told the previous destination submitter task about
- For the merkle tree processor tasks, iiuc the merkle tree that it builds is an `Arc<RwLock<MerkleTreeBuilder>>` from `self.prover_syncs[origin]` - if we respawn, then it'll have already built out a merkle tree in the first task, and then when it restarts I think it'll keep building upon that existing merkle tree, but it'll do its loop of inserting each message into the merkle tree starting from index 0 again. So I think it may break the merkle tree that's built by double inserting messages into it

Reply:
> I thought we settled on panicking

fully in agreement, I had misinterpreted `try_join`'s signature.

Really appreciate the insights into the subtleties of how tasks tie together, definitely don't want to deal with any of that behavior for now